WRPAccessControl.go

```diff
@@ -0,0 +1,158 @@
+	"github.com/xmidt-org/bascule"
+	"github.com/xmidt-org/webpa-common/basculechecks"
+	"github.com/xmidt-org/webpa-common/xhttp"
+	"github.com/xmidt-org/wrp-go/v2"
+)
+
+//partnerAuthority errors
+var (
+	ErrTokenMissing           = &xhttp.Error{Code: http.StatusInternalServerError, Text: "No JWT Token was found in context"}
+	ErrTokenTypeMismatch      = &xhttp.Error{Code: http.StatusInternalServerError, Text: "Token must be a JWT"}
+	ErrPIDMissing             = &xhttp.Error{Code: http.StatusBadRequest, Text: "WRP PartnerIDs field must not be empty"}
+	ErrInvalidAllowedPartners = &xhttp.Error{Code: http.StatusForbidden, Text: "AllowedPartners JWT claim must be a non-empty list of strings"}
+	ErrPIDMismatch            = &xhttp.Error{Code: http.StatusForbidden, Text: "Unauthorized partners credentials in WRP message"}
+)
+
+//WRPCheckConfig drives the WRP Access control configuration when enabled
+type WRPCheckConfig struct {
+	Type string
+}
+
+// wrpAccessAuthority describes behavior for authorizing WRP messages
+// against defined access policies.
+type wrpAccessAuthority interface {
+	authorizeWRP(context.Context, *wrp.Message) (bool, error)
+}
+
+//authorizeWRP should run the scytale partnerID checks against incoming WRP messages
+//It takes a pointer to the wrp message as it may modify it in some cases. It returns
+//true if such modification was made. An error is returned in cases the validator
+//check failed and they are go-kit HTTP response error encoder friendly
+
+// wrpPartnersAuthority defines the access policy for which WRP messages
+// are authorized against the partners credentials of the message creator
+type wrpPartnersAccess struct {
+	strict                  bool
+	receivedWRPMessageCount metrics.Counter
+}
+
+func (p *wrpPartnersAccess) withFailure(labelValues ...string) metrics.Counter {
+	if !p.strict {
+		return p.withSuccess(labelValues...)
+	}
+	return p.receivedWRPMessageCount.With(append(labelValues, OutcomeLabel, Rejected)...)
+}
+
+func (p *wrpPartnersAccess) withSuccess(labelValues ...string) metrics.Counter {
+	return p.receivedWRPMessageCount.With(append(labelValues, OutcomeLabel, Accepted)...)
+}
+
+//authorizeWRP runs the partners access policy against the WRP and returns an error if the check fails.
+//When the policy is not strictly enforced,
+// Additionally, when the policy is not a boolean is returned for failure cases where the policy autocorrects the WRP contents
+func (p *wrpPartnersAccess) authorizeWRP(ctx context.Context, message *wrp.Message) (bool, error) {
+	var (
+		auth, ok    = bascule.FromContext(ctx)
+		satClientID = "none"
+	)
+
+	if !ok {
+		p.withFailure(ClientIDLabel, satClientID, ReasonLabel, TokenMissing).Add(1)
+
+		if p.strict {
+			return false, ErrTokenMissing
+		}
+		return false, nil
+	}
+
+	token := auth.Token
+
+	if token.Type() != "jwt" {
+		p.withFailure(ClientIDLabel, satClientID, ReasonLabel, TokenTypeMismatch).Add(1)
+
+		if p.strict {
+			return false, ErrTokenTypeMismatch
```

Comment:
So if we have a basic auth and we're enforcing partner IDs, we'll reject it.  Is this the way we want to handle it?